File: src/fs/fs.rs
/*
 MIT License

 Copyright (c) 2019 Max Kuznetsov <syhpoon@syhpoon.ca>

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
*/

use std::io::SeekFrom;
use std::collections::{HashMap, VecDeque};
use std::rc::Rc;
use std::cell::RefCell;

use failure::{Error, format_err};

use crate::fs;
use crate::device::Device;
use super::types::*;
use super::{Bitmap, Block, Directory, Inode, Superblock};
use super::inode::{InodeType, InodePerm};
use super::fs_info::FsInfo;

/// Filesystem structure is the main entry point to all
/// the filesystem operations
pub struct Filesystem<T: Device> {
    device: T,
    superblock: Superblock,
    inode_bitmap: Bitmap,
    data_bitmap: Bitmap,
    next_free_inode: Option<usize>,
    next_free_block: Option<usize>,
    block_cache: HashMap<BlockPtr, Rc<RefCell<Block>>>,
    inode_cache: HashMap<InodePtr, Rc<RefCell<Inode>>>,
    //dir_cache: HashMap<InodePtr, Rc<RefCell<Inode>>>,
}

impl<T: Device> Filesystem<T> {
    /// Create a new filesystem on a given device.
    /// Existing data on the device will be destroyed.
    pub fn create(mut device: T) -> Result<Self, Error> {
        // Skip boot block
        device.seek(SeekFrom::Start(fs::BOOT_BLOCK_SIZE))?;

        // Create superblock
        let superblock = Filesystem::create_superblock(&mut device)?;

        // Write inode bitmap
        let mut inode_bitmap = Bitmap::new(superblock.inode_bitmap_size());

        // Zeroth inode is not available
        inode_bitmap.set(0);
        device.write_all(inode_bitmap.as_slice())?;

        // Write data bitmap
        let mut data_bitmap = Bitmap::new(superblock.data_bitmap_size());

        // Zeroth block is not available
        data_bitmap.set(0);
        device.write_all(data_bitmap.as_slice())?;

        let mut fs = Filesystem {
            device,
            inode_bitmap,
            data_bitmap,
            superblock,
            next_free_inode: Some(1),
            next_free_block: Some(1),
            inode_cache: HashMap::new(),
            block_cache: HashMap::new(),
        };

        // We need to have a root directory on a newly created filesystem
        let mut inode = fs.allocate_inode()?;
        inode.set_dirty(true);

        inode.set_type(InodeType::Directory);
        inode.set_perm(
            InodePerm(
                InodePerm::owner_read_write_execute() |
                    InodePerm::group_read() |
                    InodePerm::group_execute() |
                    InodePerm::others_read() |
                    InodePerm::others_execute()));

        let block = fs.allocate_block()?;

        // The inode now has one data block
        inode.add_block(block.idx)?;

        // Put block into cache
        fs.put_block(Rc::new(RefCell::new(block)));

        let mut dir = Directory::load(&inode, &mut fs)?;

        dir.add_entry(String::from("."), inode.ino, &mut fs)?;
        dir.add_entry(String::from(".."), inode.ino, &mut fs)?;

        fs.put_inode(Rc::new(RefCell::new(inode)));

        //TODO: Add dir to cache?

        // Flush changes to device
        fs.flush()?;

        Ok(fs)
    }

    /// Load existing filesystem from a device
    pub fn load(mut device: T) -> Result<Self, Error> {
        // Skip boot block
        device.seek(SeekFrom::Start(fs::BOOT_BLOCK_SIZE))?;

        let mut buf: Vec<u8> = vec![0; fs::SUPERBLOCK_SIZE as usize];
        device.read(buf.as_mut_slice())?;

        let sb = Superblock::load(buf.as_slice())?;

        // Load bitmaps
        let mut inode_bitmap = Filesystem::load_bitmap(sb.inode_bitmap_size(),
                                                       &mut device)?;

        let mut data_bitmap = Filesystem::load_bitmap(sb.data_bitmap_size(),
                                                      &mut device)?;

        let next_free_inode = inode_bitmap.next_clear_bit(0);
        let next_free_block = data_bitmap.next_clear_bit(0);

        let fs = Filesystem {
            device,
            inode_bitmap,
            data_bitmap,
            next_free_inode,
            next_free_block,
            superblock: sb,
            block_cache: HashMap::new(),
            inode_cache: HashMap::new(),
        };

        Ok(fs)
    }

    pub fn info(&self) -> FsInfo {
        FsInfo {
            block_size: self.superblock.block_size,
            num_inodes: self.superblock.num_inodes,
            inode_size: self.superblock.inode_size,
            num_inode_bitmap_blocks: self.superblock.num_inode_bitmap_blocks,
            num_data_bitmap_blocks: self.superblock.num_data_bitmap_blocks,
            num_inode_blocks: self.superblock.num_inode_blocks,
            first_data_block_offset: self.superblock.first_data_block_offset,
        }
    }

    fn create_superblock(device: &mut T) -> Result<Superblock, Error> {
        let sb_params = fs::superblock::Params {
            disk_size: device.capacity()?,
            ..Default::default()
        };

        let sb = fs::Superblock::new(sb_params);

        // Write superblock
        let bytes: Vec<u8> = bincode::serialize(&sb)?;
        device.write(bytes.as_slice())?;

        Ok(sb)
    }

    fn load_bitmap(size: usize, device: &mut T) -> Result<Bitmap, Error> {
        let mut buf: Vec<u8> = vec![0; size];

        device.read(buf.as_mut_slice())?;

        Ok(Bitmap::from_buf(buf))
    }

    fn allocate_inode(&mut self) -> Result<Inode, Error> {
        let idx = match self.next_free_inode {
            Some(idx) => idx,
            None => return Err(format_err!("no free inodes left"))
        };

        // Mark inode as used
        self.inode_bitmap.set(idx);

        let inode = Inode::new(idx as u32);
        self.next_free_inode = self.inode_bitmap.next_clear_bit(idx);

        Ok(inode)
    }

    fn allocate_block(&mut self) -> Result<Block, Error> {
        let idx = match self.next_free_block {
            Some(idx) => idx,
            None => return Err(format_err!("no free blocks left"))
        };

        // Mark block as used
        self.data_bitmap.set(idx);

        let block = Block::new(idx as u32,
                               self.superblock.block_size as usize);
        self.next_free_block = self.data_bitmap.next_clear_bit(idx);

        Ok(block)
    }

    #[inline(always)]
    fn block_offset(&self, block_idx: BlockPtr) -> u64 {
        self.superblock.first_data_block_offset +
            (block_idx * self.superblock.block_size) as u64
    }

    #[inline(always)]
    fn inode_offset(&self, inode_idx: InodePtr) -> u64 {
        self.superblock.first_inode_block_offset +
            (inode_idx * self.superblock.inode_size) as u64
    }

    /// Put block into a cache
    fn put_block(&mut self, block: Rc<RefCell<Block>>) {
        let idx = block.borrow().idx;
        self.block_cache.insert(idx, block);
    }

    /// Put inode into a cache
    fn put_inode(&mut self, inode: Rc<RefCell<Inode>>) {
        let idx = inode.borrow().ino;
        self.inode_cache.insert(idx, inode);
    }

    /// Return a given block.
    /// The block will either be returned from a cache or loaded from device
    pub fn get_block(&mut self, idx: BlockPtr) -> Result<
        Rc<RefCell<Block>>, Error> {
        match self.block_cache.get_mut(&idx) {
            Some(block) => Ok(Rc::clone(&block)),
            None => {
                println!("get_block ---------- idx={}", idx);
                // Load from device
                let mut buf: Vec<u8> = vec![
                    0; self.superblock.block_size as usize];

                self.device.seek(SeekFrom::Start(self.block_offset(idx)))?;
                self.device.read(buf.as_mut_slice())?;

                let block = Block::from_buf(idx, buf);

                let brc = Rc::new(RefCell::new(block));
                let res = Rc::clone(&brc);

                self.put_block(brc);

                Ok(res)
            }
        }
    }

    fn flush(&mut self) -> Result<(), Error> {
        let mut blocks: VecDeque<Rc<RefCell<Block>>> = VecDeque::new();
        let mut inodes: VecDeque<Rc<RefCell<Inode>>> = VecDeque::new();

        for cell in self.block_cache.values() {
            if cell.borrow().is_dirty() {
                blocks.push_back(Rc::clone(&cell));
            }
        }

        for cell in self.inode_cache.values() {
            if cell.borrow().is_dirty() {
                inodes.push_back(Rc::clone(&cell));
            }

            //TODO: inode blocks?
        }

        for cell in blocks {
            let mut block = cell.borrow_mut();
            self.flush_block(&block)?;
            block.set_dirty(false);
        };

        for cell in inodes {
            let mut inode = cell.borrow_mut();
            self.flush_inode(&inode)?;
            inode.set_dirty(false);
        };

        Ok(())
    }

    fn flush_block(&mut self, block: &Block) -> Result<(), Error> {
        self.device.seek(SeekFrom::Start(self.block_offset(block.idx)))?;
        self.device.write(block.data.as_slice())?;

        Ok(())
    }

    fn flush_inode(&mut self, inode: &Inode) -> Result<(), Error> {
        let bytes: Vec<u8> = bincode::serialize(&inode)?;
        self.device.seek(SeekFrom::Start(self.inode_offset(inode.ino)))?;

        (&mut self.device).write(bytes.as_slice())?;

        Ok(())
    }
}

File: pkg/fs/fs.go
package fs

import (
	"context"
	"fmt"
	"sync"
	"syscall"

	"github.com/syhpoon/learnfs/pkg/device"
)

type Filesystem struct {
	dev            device.Device
	superblock     *Superblock
	inodeAllocator InodeAllocator
	blockAllocator BlockAllocator
	blockCache     *blockCache
	inodeCache     *InodeCache
	dirCache       *dirCache
	pool           *bufPool
	flusher        *flusher
	blockSize      uint32
}

// Create a new filesystem on a given device
func Create(dev device.Device) error {
	// Create superblock instance
	sbParams := DefaultSuperblockParams()
	sbParams.DiskSize = dev.Capacity()

	pool := newBufPool(sbParams.BlockSize)
	sb := NewSuperblock(sbParams)

	// Write metadata
	if err := writeFsMetadata(pool, sb, dev); err != nil {
		return fmt.Errorf("failed to write metadata: %w", err)
	}

	inodeBitmapBuf := make(Buf, sb.InodeBitmapSize())
	inodeBitmap := newBitmap(sb.NumInodes, inodeBitmapBuf)
	inodeBitmap.set(0)

	dataBitmapBuf := make(Buf, sb.DataBitmapSize())
	dataBitmap := newBitmap(sb.NumDataBlocks, dataBitmapBuf)
	dataBitmap.set(0)

	inodeAllocator := newInodeAllocatorSimple(inodeBitmap)
	blockAllocator := newBlockAllocatorSimple(dataBitmap, sb.BlockSize)
	blockCache := newBlockCache(dev, blockAllocator, pool, sb)
	inodeCache := NewInodeCache(dev, inodeAllocator, pool, sb)
	dirCache := newDirCache(pool, inodeCache, blockCache, sb, dev)
	flusherObj := &flusher{
		pool:           pool,
		blockCache:     blockCache,
		inodeCache:     inodeCache,
		dirCache:       dirCache,
		sb:             sb,
		inodeAllocator: inodeAllocator,
		blockAllocator: blockAllocator,
		dev:            dev,
	}

	fs := &Filesystem{
		dev:            dev,
		superblock:     sb,
		inodeAllocator: inodeAllocator,
		blockAllocator: blockAllocator,
		pool:           pool,
		blockCache:     blockCache,
		inodeCache:     inodeCache,
		dirCache:       dirCache,
		blockSize:      sb.BlockSize,
		flusher:        flusherObj,
	}

	// Create a root directory
	if err := fs.createRootDir(); err != nil {
		return fmt.Errorf("failed to create root directory: %w", err)
	}

	return nil
}

// Load filesystem from a device
func Load(dev device.Device) (*Filesystem, error) {
	offset := uint64(0)

	// Load superblock
	buf := make(Buf, SUPERBLOCK_SIZE)
	op := &device.Op{
		Buf:    buf,
		Offset: offset,
	}

	if err := dev.Read(op); err != nil {
		return nil, fmt.Errorf("failed to read superblock: %w", err)
	}

	sb, err := LoadSuperblock(buf)
	if err != nil {
		return nil, fmt.Errorf("failed to load superblock: %w", err)
	}

	offset += uint64(SUPERBLOCK_SIZE)

	// Load bitmaps
	inodeBitmapBuf := make(Buf, sb.InodeBitmapSize())
	ibmOp := &device.Op{
		Buf:    inodeBitmapBuf,
		Offset: offset,
	}

	dataBitmapBuf := make(Buf, sb.DataBitmapSize())
	dbmOp := &device.Op{
		Buf:    dataBitmapBuf,
		Offset: offset + uint64(sb.InodeBitmapSize()),
	}

	if err := dev.Read(ibmOp, dbmOp); err != nil {
		return nil, fmt.Errorf("failed to read bitmaps: %w", err)
	}

	inodeBitmap := newBitmap(sb.NumInodes, inodeBitmapBuf)
	dataBitmap := newBitmap(sb.NumDataBlocks, dataBitmapBuf)
	pool := newBufPool(sb.BlockSize)

	inodeAllocator := newInodeAllocatorSimple(inodeBitmap)
	blockAllocator := newBlockAllocatorSimple(dataBitmap, sb.BlockSize)
	blockCache := newBlockCache(dev, blockAllocator, pool, sb)
	inodeCache := NewInodeCache(dev, inodeAllocator, pool, sb)
	dirCache := newDirCache(pool, inodeCache, blockCache, sb, dev)
	flusherObj := &flusher{
		pool:           pool,
		blockCache:     blockCache,
		inodeCache:     inodeCache,
		dirCache:       dirCache,
		sb:             sb,
		inodeAllocator: inodeAllocator,
		blockAllocator: blockAllocator,
		dev:            dev,
	}

	return &Filesystem{
		dev:            dev,
		superblock:     sb,
		inodeAllocator: inodeAllocator,
		blockAllocator: blockAllocator,
		pool:           pool,
		blockCache:     blockCache,
		inodeCache:     inodeCache,
		dirCache:       dirCache,
		flusher:        flusherObj,
		blockSize:      sb.BlockSize,
	}, nil
}

func (fs *Filesystem) RunFlusher(ctx context.Context, wg *sync.WaitGroup) {
	fs.flusher.run(ctx, wg)
}

func (fs *Filesystem) Shutdown() error {
	return fs.dev.Close()
}

func (fs *Filesystem) BlockSize() uint32 {
	return fs.superblock.BlockSize
}

func (fs *Filesystem) Superblock() *Superblock {
	return fs.superblock
}

func (fs *Filesystem) createRootDir() error {
	ino, err := fs.inodeAllocator.AllocateInode()
	if err != nil {
		return fmt.Errorf("failed to allocate inode: %w", err)
	}

	ino.Mode = syscall.S_IFDIR | 0777

	block, err := fs.blockAllocator.AllocateBlock()
	if err != nil {
		return fmt.Errorf("failed to allocate block: %w", err)
	}

	if err := ino.AddBlockPtr(block.ptr); err != nil {
		return fmt.Errorf("failed to add block to inode: %w", err)
	}

	dir := NewDirectory(ino, fs.blockCache, fs.pool, fs.blockSize)

	if err := dir.AddEntry(".", ino.Ptr()); err != nil {
		return fmt.Errorf("failed to add directory entry: %w", err)
	}

	if err := dir.AddEntry("..", ino.Ptr()); err != nil {
		return fmt.Errorf("failed to add directory entry: %w", err)
	}

	ino.Nlink = 2
	ino.SetDirty(true)
	fs.inodeCache.AddInode(ino)

	block.setDirty(true)
	fs.blockCache.addBlock(block)

	fs.dirCache.addDir(dir)

	fs.flusher.flushAll()
	fs.flusher.flushBitmaps()

	return nil
}

func writeFsMetadata(pool *bufPool, sb *Superblock, dev device.Device) error {
	inodeBitmap := newBitmap(sb.NumInodes, make([]byte, sb.InodeBitmapSize()))
	inodeBitmap.set(0)

	dataBitmap := newBitmap(sb.NumDataBlocks, make([]byte, sb.DataBitmapSize()))
	dataBitmap.set(0)

	bin, err := serialize(pool, sb, inodeBitmap.buf, dataBitmap.buf)
	if err != nil {
		return fmt.Errorf("failed to serialize metadata: %w", err)
	}

	op := &device.Op{
		Buf:    bin,
		Offset: 0,
	}

	if err := dev.Write(op); err != nil {
		return fmt.Errorf("failed to write metadata blocks: %w", err)
	}

	return nil
}

func (fs *Filesystem) blockOffset(offset int64) (int64, int64) {
	// Get the logical block for the offset
	blockIdx := offset / int64(fs.blockSize)
	// Get the offset within the block
	blockOffset := offset % int64(fs.blockSize)

	return blockIdx, blockOffset
}

func (fs *Filesystem) getInodeBlock(ino *Inode, idx int64) (BlockPtr, error) {
	if err := fs.prepareInodeBlock(ino, idx); err != nil {
		return 0, fmt.Errorf("failed to prepare inode block: %w", err)
	}

	return ino.blocks[idx], nil
}

func (fs *Filesystem) setInodeBlock(ino *Inode, idx int64, ptr BlockPtr) error {
	if err := fs.prepareInodeBlock(ino, idx); err != nil {
		return fmt.Errorf("failed to prepare inode block: %w", err)
	}

	if idx >= DIRECT_BLOCKS_NUM {
		ptrsPerPage := int64(fs.blockSize) / BlockPtrSize
		singleMaxBlocks := int64(DIRECT_BLOCKS_NUM) + ptrsPerPage

		var blk *block
		var s int64
		var err error

		if idx < singleMaxBlocks {
			blk, err = fs.blockCache.getBlock(ino.IndirectBlock)
			if err != nil {
				return fmt.Errorf("failed to get indirect block %d for inode %d: %w",
					ino.IndirectBlock, ino.ptr, err)
			}

			s = (idx - DIRECT_BLOCKS_NUM) * BlockPtrSize
		} else {
			doublePtr := ino.doubleBlocks[idx/int64(fs.blockSize)]
			blk, err = fs.blockCache.getBlock(doublePtr)
			if err != nil {
				return fmt.Errorf("failed to get double indirect block %d for inode %d: %w",
					ino.IndirectBlock, ino.ptr, err)
			}

			s = (idx - singleMaxBlocks) * BlockPtrSize
		}

		blk.Lock()
		binEncoding.PutUint32(blk.data[s:s+BlockPtrSize], ptr)
		blk.Unlock()
		blk.setDirty(true)
	}

	ino.blocks[idx] = ptr
	ino.SetDirty(true)

	return nil
}

func (fs *Filesystem) prepareInodeBlock(ino *Inode, idx int64) error {
	// Block already present (either direct or loaded previously)
	if idx < int64(len(ino.blocks)) {
		return nil
	}

	ptrsPerPage := int64(fs.blockSize) / BlockPtrSize

	// Number of single indirect pointers
	singleMaxBlocks := int64(DIRECT_BLOCKS_NUM) + ptrsPerPage

	// Number of double indirect pointers
	doubleMaxBlocks := singleMaxBlocks + ptrsPerPage*ptrsPerPage

	if idx < singleMaxBlocks {
		// Need to allocate the indirect block pointer
		if ino.IndirectBlock == 0 {
			if err := fs.allocateSingleIndirectBlock(ino, ptrsPerPage); err != nil {
				return err
			}
		} else {
			// Load block pointers
			blocks, err := fs.loadIndirectBlocks(ino.IndirectBlock, ptrsPerPage)
			if err != nil {
				return err
			}

			ino.addBlocksNoLock(blocks)
		}

		ino.SetDirty(true)
	} else if idx < doubleMaxBlocks {
		var doubleBlk *block
		var err error

		// Double indirect
		if ino.DoubleIndirectBlock == 0 {
			doubleBlk, err = fs.allocateDoubleIndirectBlock(ino)
			if err != nil {
				return err
			}
		} else {
			doubleBlk, err = fs.blockCache.getBlock(ino.DoubleIndirectBlock)
			if err != nil {
				return fmt.Errorf("failed to load double indirect block %d: %w",
					ino.DoubleIndirectBlock, err)
			}
		}

		doubleBlk.Lock()
		defer doubleBlk.Unlock()

	LOOP:
		for l1 := int64(0); l1 < ptrsPerPage; l1++ {
			var l1Block *block
			l1Offset := l1 * BlockPtrSize
			l1BlockPtr := binEncoding.Uint32(doubleBlk.data[l1Offset : l1Offset+BlockPtrSize])

			if l1BlockPtr == 0 {
				l1Block, err = fs.allocateL1DoubleIndirectBlock(doubleBlk, l1Offset)
				if err != nil {
					return err
				}
			} else {
				l1Block, err = fs.blockCache.getBlock(l1BlockPtr)
				if err != nil {
					return fmt.Errorf("failed to load l1 double indirect block %d: %w",
						l1BlockPtr, err)
				}
			}

			l1Block.Lock()
			defer l1Block.Unlock()

			for l2 := int64(0); l2 < ptrsPerPage; l2++ {
				l2Offset := l2 * BlockPtrSize
				l2BlockPtr := binEncoding.Uint32(l1Block.data[l2Offset : l2Offset+BlockPtrSize])

				if l2BlockPtr == 0 {
					blk, err := fs.blockAllocator.AllocateBlock()
					if err != nil {
						return fmt.Errorf(
							"failed to allocate l2 double indirect block for inode %d: %w",
							ino.ptr, err)
					}

					fs.blockCache.addBlock(blk)
					binEncoding.PutUint32(l1Block.data[l2Offset:l2Offset+BlockPtrSize], blk.ptr)
					l1Block.setDirty(true)

					// Since it's a newly allocated indirect block, all the pointers are
					// unused (0) at this point
					ino.addBlocksNoLock(make([]BlockPtr, ptrsPerPage))
					ino.doubleBlocks[idx/int64(fs.blockSize)] = blk.ptr
				} else {
					blocks, err := fs.loadIndirectBlocks(l2BlockPtr, ptrsPerPage)
					if err != nil {
						return err
					}

					ino.addBlocksNoLock(blocks)
				}

				if int64(len(ino.blocks)) > idx {
					break LOOP
				}
			}
		}
	} else {
		return fmt.Errorf("inode %d block index exceeding maximum file size: %d", ino.ptr, idx)
	}

	return nil
}

func (fs *Filesystem) allocateSingleIndirectBlock(ino *Inode, ptrsPerPage int64) error {
	blk, err := fs.blockAllocator.AllocateBlock()
	if err != nil {
		return fmt.Errorf(
			"failed to allocate indirect block for inode %d: %w", ino.ptr, err)
	}

	fs.blockCache.addBlock(blk)

	// Since it's a newly allocated indirect block, all the pointers are
	// unused (0) at this point
	ino.blocks = append(ino.blocks, make([]BlockPtr, ptrsPerPage)...)
	ino.IndirectBlock = blk.ptr

	return nil
}

func (fs *Filesystem) allocateDoubleIndirectBlock(ino *Inode) (*block, error) {
	blk, err := fs.blockAllocator.AllocateBlock()
	if err != nil {
		return nil, fmt.Errorf(
			"failed to allocate double indirect block for inode %d: %w", ino.ptr, err)
	}

	fs.blockCache.addBlock(blk)
	ino.DoubleIndirectBlock = blk.ptr

	return blk, nil
}

func (fs *Filesystem) allocateL1DoubleIndirectBlock(
	doubleBlk *block, offset int64) (*block, error) {

	blk, err := fs.blockAllocator.AllocateBlock()
	if err != nil {
		return nil, fmt.Errorf("failed to allocate l1 double indirect block: %w", err)
	}

	fs.blockCache.addBlock(blk)
	binEncoding.PutUint32(doubleBlk.data[offset:offset+BlockPtrSize], blk.ptr)
	doubleBlk.setDirty(true)

	return blk, nil
}

func (fs *Filesystem) loadIndirectBlocks(ptr BlockPtr, ptrsPerPage int64) ([]BlockPtr, error) {
	blk, err := fs.blockCache.getBlock(ptr)
	if err != nil {
		return nil, fmt.Errorf("failed to load indirect block %d: %w", ptr, err)
	}

	blocks := make([]BlockPtr, ptrsPerPage)

	for i := range blocks {
		s := int64(i) * BlockPtrSize
		blocks[i] = binEncoding.Uint32(blk.data[s : s+BlockPtrSize])
	}

	return blocks, nil
}
